### DISCUSSION TITLE
Any reason that an order canceled still keep the `completed_at` value?

### DISCUSSION THREAD

**Discussion by jtapia** [solidusio/solidus#4800] [Troubleshooting]
I'm seeing that after order cancellation, the value `completed_at` is still populated, any reason for this? or a canceled order is considered as `completed`?

**Comment by seandawes** [comment]
completed_at is not what really defines the order state of complete. order state of complete would. completed_at is used to determine historical sales and other reporting as created_at and completed_at are often times lengthy date diff

**Reply by jarednorman** [reply]
Sean has this correct. You need to keep a record of when the order was completed for the purposes of sales reports, accounting, etc.

I do recognize that the whole `complete?` versus `completed?` thing is confusing.

- `Spree::Order#complete?` returns true if the order is currently in the `complete` state. If the order has been cancelled it will return false.
- `Spree::Order#completed?` returns true if the order ever got through checkout by checking if the `completed_at` timestamp is set.

While the naming is confusing, they are both useful in different contexts.

**Comment by jtapia** [comment]
Thanks for the replies and the explaining

Not sure if the concept of `complete` order makes sense for this case, for example:

```
irb(main):011:0> user
=> #<Spree::User ...>
irb(main):012:0> user.orders
=> #<ActiveRecord::Associations::CollectionProxy [#<Spree::Order ...]>
irb(main):013:0> user.orders.count
   (1.3ms)  SELECT COUNT(*) FROM "spree_orders" WHERE "spree_orders"."user_id" = $1  [["user_id", 11111]]
=> 3
irb(main):014:0> user.orders.map(&:state)
=> ["cart", "canceled", "canceled"]
irb(main):015:0> user.orders.complete
  Spree::Order Load (5.1ms)  SELECT "spree_orders".* FROM "spree_orders" WHERE "spree_orders"."user_id" = $1 AND ("spree_orders"."completed_at" IS NOT NULL)  [["user_id", 11111]]
=> #<ActiveRecord::AssociationRelation [#<Spree::Order id: ...]>
irb(main):016:0> user.orders.complete.count
   (1.3ms)  SELECT COUNT(*) FROM "spree_orders" WHERE "spree_orders"."user_id" = $1 AND ("spree_orders"."completed_at" IS NOT NULL)  [["user_id", 11111]]
=> 2
irb(main):017:0> user.orders.complete.map(&:state)
  Spree::Order Load (1.4ms)  SELECT "spree_orders".* FROM "spree_orders" WHERE "spree_orders"."user_id" = $1 AND ("spree_orders"."completed_at" IS NOT NULL)  [["user_id", 11111]]
=> ["canceled", "canceled"]
```

it's considering the `canceled` orders as `completed`, I mean, I would expect to see only orders on `complete` state

**Reply by seandawes** [reply]
You would not want to drop completed_at info as this will impact all forms of reporting. Similar confusing naming conventions occur with order numbers even as a cart has an order number but you would think an order number should only exist for an actual order. This is where the state machine really comes into play and you just need to look at everything around states. 

Added complexity comes into play when you have multiple stock locations and multiple line items.

**Reply by jarednorman** [reply]
It's true that our naming here is inconsistent. The `Spree::Order.complete` scope returns orders that would `Spree::Order#completed?` would return true for, not `Spree::Order#complete?`.

`Spree::Order.by_state("complete")` will give you what you need if you need to exclude cancelled orders.
